Sources/NIO/Bootstrap.swift

```diff
@@ -387,9 +393,16 @@ public final class ClientBootstrap {
             channelInitializer(channel).then {
                 channelOptions.applyAll(channel: channel)
             }.then {
-                channel.register()
-            }.then {
-                body(channel)
+                // this is pretty delicate at the moment:
+                // `body` (which will `connect`) must be _synchronously_ follow `register`, otherwise in our current
+                // implementation, `epoll` will send us `EPOLLHUP`. To have it run synchronously, we need to invoke the
+                // `then` on the eventloop that the
+                // `register` will succeed.
+                assert(channel.eventLoop.inEventLoop)
+                return channel.register().then {
+                    assert(channel.eventLoop.inEventLoop)
+                    return body(channel)
+                }
```

Comment:
@weissi consider making a utility function for the register / body execution and also use it for register / bind above. Better to keep all the "special" code in one place